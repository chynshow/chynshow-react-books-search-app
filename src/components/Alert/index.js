import React, { useContext } from 'react';
import { AlertContext } from '../../state/AlertContext';

const Alert = () => {
  // Get data from global app state
  const { message, style } = useContext(AlertContext);

  return (
    <>
      {message && <div className={`c-alert c-aler--${style}`}>{message}</div>}
    </>
  );
};

export default Alert;
