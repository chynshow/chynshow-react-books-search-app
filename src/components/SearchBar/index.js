import React, { useState, useContext } from 'react';
import Input from './../Input';
import { AppContext } from '../../state/AppContext';
import { AlertContext } from '../../state/AlertContext';
import Settings from './../Settings';
import CogSVG from '../SVG/CogSVG';
import Overlay from './../Overlay';

const SearchBar = () => {
  return (
    <div className='l-search-bar'>
      <MainTitle />
      <div className='l-search-bar-container'>
        <SettingsContainer />
        <SearchForm />
      </div>
      <Overlay opacity='.3' background='#202124' />
    </div>
  );
};

export default SearchBar;

const SearchForm = () => {
  // Get data from global app state
  const {
    fetchBooks,
    startIndex,
    searchParams,
    maxResults,
    serSearchParams,
    clearState,
  } = useContext(AppContext);
  const { showAlert } = useContext(AlertContext);

  // Use local state for form handling
  const [values, setValues] = useState(searchParams);

  // Handle onSubmit event
  const handlerOnSubmit = (e) => {
    e.preventDefault();
    clearState();

    // Validation based on non empty fields
    if (!(Object.values(values).filter((v) => v !== '').length > 0))
      return showAlert(
        'Please add at least one search parameter!',
        'warning',
        5000
      );

    fetchBooks(values, startIndex, maxResults);
    serSearchParams(values);

    //  Clear inputs
    //  setValues({
    //  query: '',
    //  inauthor: '',
    //  inpublisher: '',
    //  isbn: '',
    //  });
  };

  return (
    <form className='c-search-form' onSubmit={handlerOnSubmit}>
      <Input
        className='c-search-form__input'
        name='query'
        value={values.query}
        placeholder='Your search query'
        onChange={(e) =>
          setValues({ ...values, [e.target.name]: e.target.value })
        }
      />
      <Input
        className='c-search-form__input'
        name='inauthor'
        value={values.inauthor}
        placeholder='Author'
        onChange={(e) =>
          setValues({ ...values, [e.target.name]: e.target.value })
        }
      />
      <Input
        className='c-search-form__input'
        name='inpublisher'
        value={values.inpublisher}
        placeholder='Publisher'
        onChange={(e) =>
          setValues({ ...values, [e.target.name]: e.target.value })
        }
      />
      <Input
        className='c-search-form__input'
        name='isbn'
        type='number'
        value={values.isbn}
        placeholder='ISBN'
        onChange={(e) =>
          setValues({ ...values, [e.target.name]: e.target.value })
        }
      />

      <button className='c-btn c-search-form__btn'>Search</button>
    </form>
  );
};

const SettingsContainer = () => {
  const [showSettings, setShowSettings] = useState(false);
  return (
    <>
      {showSettings && <Settings setShowSettings={setShowSettings} />}
      <button
        className='c-btn c-settings__btn'
        onClick={() => setShowSettings(!showSettings)}
      >
        <CogSVG />
      </button>
    </>
  );
};

const MainTitle = () => (
  <h3 className='c-main-title'>Which book you want to read next?</h3>
);
