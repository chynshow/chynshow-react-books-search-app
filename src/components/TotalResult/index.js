import React, { useContext } from 'react';
import { AppContext } from '../../state/AppContext';

const TotalResult = () => {
  // Get data from global app state
  const { totalResults, loading } = useContext(AppContext);
  return (
    <>
      <div>
        {!loading && totalResults !== null && (
          <span>
            We found{' '}
            {totalResults > 1
              ? `${totalResults} results for you!`
              : `${totalResults} result for you!`}
          </span>
        )}

        {totalResults === null && (
          <span>You don't have any books yet! Start search!</span>
        )}
      </div>
    </>
  );
};

export default TotalResult;
