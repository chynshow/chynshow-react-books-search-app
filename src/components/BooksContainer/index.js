import React, { useCallback, useContext, useRef } from 'react';
import { AppContext } from '../../state/AppContext';
import BookItem from '../BookItem';
import TotalResult from '../TotalResult';

const BooksContainer = () => {
  const {
    books,
    loading,
    startIndex,
    maxResults,
    totalResults,
    fetchBooks,
    searchParams,
    setStartIndex,
  } = useContext(AppContext);

  // Load new items based on position of last item in the list and "totalResults" parameter
  const observer = useRef();
  const lastBookElRef = useCallback(
    (el) => {
      if (loading) return;
      if (observer.current) observer.current.disconnect();
      observer.current = new IntersectionObserver((entries) => {
        if (
          entries[0].isIntersecting &&
          startIndex + maxResults < totalResults
        ) {
          setStartIndex();
          fetchBooks(searchParams, startIndex, maxResults);
        }
      });
      if (el) observer.current.observe(el);
    },
    [
      loading,
      searchParams,
      startIndex,
      maxResults,
      totalResults,
      setStartIndex,
      fetchBooks,
    ]
  );

  return (
    <ul className='l-book-container'>
      <TotalResult />
      {
        // eslint-disable-next-line array-callback-return
        books.map(({ volumeInfo, id }, idx) => {
          if (volumeInfo && volumeInfo.imageLinks) {
            if (books.length === idx + 1) {
              return (
                <BookItem
                  title={volumeInfo.title}
                  imageLinks={volumeInfo.imageLinks}
                  description={volumeInfo.description}
                  key={id}
                  reference={lastBookElRef}
                />
              );
            } else {
              return (
                <BookItem
                  key={id}
                  title={volumeInfo.title}
                  imageLinks={volumeInfo.imageLinks}
                  description={volumeInfo.description}
                />
              );
            }
          }
        })
      }
      {loading && <div className='c-loader'>Loading...</div>}
    </ul>
  );
};

export default BooksContainer;
