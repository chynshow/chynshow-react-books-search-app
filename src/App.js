import React from 'react';
import Alert from './components/Alert';
import BooksContainer from './components/BooksContainer';
import SearchBar from './components/SearchBar';
import TotalResult from './components/TotalResult';

const App = () => {
  return (
    <div className='l-app-container'>
      <Alert />
      <SearchBar />
      <TotalResult />
      <BooksContainer />
    </div>
  );
};

export default App;
