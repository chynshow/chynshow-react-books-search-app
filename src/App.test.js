import React from 'react';
import App from './App';
import { shallow } from 'enzyme';
import Alert from './components/Alert';
import BooksContainer from './components/BooksContainer';
import SearchBar from './components/SearchBar';
import TotalResult from './components/TotalResult';

it('renders without crashing', () => {
  shallow(<App />);
});

it('renders Alert component', () => {
  const wrapper = shallow(<App />);
  const c = <Alert />;
  expect(wrapper.contains(c)).toEqual(true);
});

it('renders BooksContainer component', () => {
  const wrapper = shallow(<App />);
  const c = <BooksContainer />;
  expect(wrapper.contains(c)).toEqual(true);
});

it('renders SearchBar component', () => {
  const wrapper = shallow(<App />);
  const c = <SearchBar />;
  expect(wrapper.contains(c)).toEqual(true);
});

it('renders TotalResult component', () => {
  const wrapper = shallow(<App />);
  const c = <TotalResult />;
  expect(wrapper.contains(c)).toEqual(true);
});
