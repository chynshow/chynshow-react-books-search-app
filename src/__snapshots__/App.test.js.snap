// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render App component 1`] = `
<div
  className="l-app-container"
>
  <Alert />
  <SearchBar />
  <TotalResult />
  <BooksContainer />
</div>
`;
